src/View/InstallQuickInput.ts

```diff
@@ -18,95 +18,160 @@ import * as vscode from 'vscode';
+    return this.toolchainEnv as ToolchainEnv;
+  }
+
+  getToolchainType(): string {
+    if (this.error !== undefined || this.toolchainType === undefined) {
+      throw new Error('wrong calling');
+    }
+    return this.toolchainType as string;
+  }
+
+  getToolchain(): Toolchain {
+    if (this.error !== undefined || this.toolchain === undefined) {
+      throw new Error('wrong calling');
+    }
+    return this.toolchain as Toolchain;
+  }
+
+  getVersion(): string {
+    if (this.error !== undefined || this.version === undefined) {
+      throw new Error('wrong calling');
+    }
+    return this.version as string;
+  }
+
+  getError(): string|undefined {
+    return this.error;
+  }
+
+  getQuickPickItems(items: string[]): vscode.QuickPickItem[] {
+    return items.map(label => ({label}));
   }
 
-  async function collectInputs() {
-    const state = {} as Partial<State>;
-    await MultiStepInput.run(input => pickBackend(input, state));
-    return state as State;
+  getAllToolchainEnvNames(): string[] {
+    return Object.keys(gToolchainEnvMap);
   }
 
-  const title = 'Choose Compiler Toolchain';
+  getToolchainEnvFromGlobal(key: string): ToolchainEnv {
+    return gToolchainEnvMap[key];
+  }
 
-  class InnerButton implements vscode.QuickInputButton {
-    constructor(public iconPath: vscode.ThemeIcon, public tooltip: string) {}
+  getToolchainTypes(): string[] {
+    return this.toolchainEnv!.getToolchainTypes();
   }
 
-  const updateButton = new InnerButton(new vscode.ThemeIcon('refresh'), 'Update version list');
+  getVersions(toolchains: Toolchain[]): string[] {
+    return toolchains.map(
+        (value) => value.info.version !== undefined ? value.info.version.str() : '');
+  }
 
-  async function pickBackend(input: MultiStepInput, state: Partial<State>) {
-    const backendGroups: vscode.QuickPickItem[] =
-        Object.keys(gToolchainEnvMap).map((label) => ({label}));
-    state.backend = await input.showQuickPick({
+  async pickBackend(input: MultiStepInput, state: Partial<State>) {
+    const backendGroups = this.getQuickPickItems(this.getAllToolchainEnvNames());
+    state.selectedItem = await input.showQuickPick({
       title,
       step: 1,
       totalSteps: 3,
       placeholder: 'Pick toolchain backend',
       items: backendGroups,
       shouldResume: shouldResume
     });
-    state.toolchainEnv = gToolchainEnvMap[state.backend.label];
-    return (input: MultiStepInput) => pickType(input, state);
+    this.toolchainEnv = this.getToolchainEnvFromGlobal(state.selectedItem.label);
   }
 
-  async function pickType(input: MultiStepInput, state: Partial<State>) {
-    if (state.toolchainEnv === undefined) {
+  async pickType(input: MultiStepInput, state: Partial<State>) {
+    if (this.toolchainEnv === undefined) {
+      this.error = 'Backend is undefined';
       throw Error('Backend is undefined');
     }
-    const types = state.toolchainEnv.getToolchainTypes();
+
+    const types = this.getToolchainTypes();
```

Comment:
About `getToolchainTypes()` function:

The existing flow already checked whether the `toolchainEnv` is undefined or not. However, in this case, this `getToolchainTypes()` function can be called alone, so it seems that there should be a code in `getToolchainTypes()` function to check whether toolchainEnv is valid.

Reply:
Great point.